### ISSUE TITLE
Multitenancy: New RunnerDeployment not using "githubEnterpriseServerURL" value

### ISSUE THREAD

**Issue by Jalmeida1994** [actions-runner-controller/actions-runner-controller#1714]
### Controller Version

canary

### Helm Chart Version

0.20.2 (main branch)

### CertManager Version

1.9

### Deployment Method

Helm

### cert-manager installation

Followed every procedure.

### Checks

- [X] This isn't a question or user support case (For Q&A and community support, go to [Discussions](https://github.com/actions-runner-controller/actions-runner-controller/discussions). It might also be a good idea to contract with any of contributors and maintainers if your business is so critical and therefore you need priority support
- [X] I've read [releasenotes](https://github.com/actions-runner-controller/actions-runner-controller/tree/master/docs/releasenotes) before submitting this issue and I'm sure it's not due to any recently-introduced backward-incompatible changes
- [X] My actions-runner-controller version (v0.x.y) does support the feature
- [X] I've already upgraded ARC (including the CRDs, see charts/actions-runner-controller/docs/UPGRADING.md for details) to the latest and it didn't fix the issue

### Resource Definitions

```yaml
---
# Source: secret.yaml
apiVersion: v1
kind: Secret
metadata:
  name: organization-github-app-secret  
  namespace: organization
data:
  github_app_id: ***
  github_app_installation_id: ***
  github_app_private_key: ***
---
# Source: hra.yaml
apiVersion: actions.summerwind.dev/v1alpha1
kind: HorizontalRunnerAutoscaler
metadata:
  name: organization-autoscaler
  namespace: organization
spec:
  githubAPICredentialsFrom:
    secretRef: 
      name: organization-github-app-secret
  scaleDownDelaySecondsAfterScaleOut: 300
  scaleTargetRef:
    name: organization-runners
  minReplicas: 0
  maxReplicas: 5
  scaleUpTriggers:
  - githubEvent: {}
    amount: 1
    duration: 30m
---
# Source: runner_deployment.yaml
apiVersion: actions.summerwind.dev/v1alpha1
kind: RunnerDeployment
metadata:
  name: organization-runners
  namespace: organization
spec:
  template:
    metadata:
    spec:
      githubAPICredentialsFrom:
        secretRef: 
          name: organization-github-app-secret 
      ephemeral: true
      labels:
        - organization-runners
      organization: organization
      containers:
        - name: runner
          imagePullPolicy: IfNotPresent
```

### To Reproduce

```markdown
1. Install actions-runner-controller chart from ´main´ branch with controller tag set to 'canary'. Added a default auth secret "controller-manager" to it;
2. Deploy RunnerDeployment, HRA and Secret with App Auth in new namespace;
3. Runner can't register with GitHub Enterprise Server.
```

### Describe the bug

New `RunnerDeployment` can't register with GitHub Enterprise Server when using a new Secret for `githubAPICredentialsFrom` (multi tenancy) with a different GitHub Auth. Controller outputs:

```shell
2022-08-12T09:18:59Z	ERROR	actions-runner-controller.runner	Failed to get new registration token	{"runner": "organization-runners-dxs7p-2qnz6", "error": "failed to create registration token: Post \"https://api.github.com/orgs/organization/actions/runners/registration-token\": could not refresh installation id ***'s token: received non 2xx response status \"401 Unauthorized\" when fetching https://api.github.com/app/installations/***/access_tokens"}
```

From what I'm seeing it doesn't seem to be using the `githubEnterpriseServerURL` specified in the controller to authenticate, and yes the Cloud one: **https://api.github.com/** instead of **https://HOSTNAME/api/v3/**.

However if I deploy a `RunnerDeployment` in the same namespace as the controller and without the `githubAPICredentialsFrom` (using the default "controller-manager" secret) it uses the correct `githubEnterpriseServerURL`.

### Describe the expected behavior

To register successfully with GitHub Enterprise Server by using the correct URL specified in the controller key `githubEnterpriseServerURL`.

### Controller Logs

```shell
https://gist.github.com/Jalmeida1994/448c18fb6fa99c69e0e9fae68e0d47de
```

### Runner Pod Logs

```shell
N/A
```

### Additional Context

Well I hope this time I found a real issue and not just me being dumb ahah. 

Cheers!

**Comment by mumoshu** [comment]
@Jalmeida1994 Ahhh, good catch! Thanks a lot for reporting.

This does seem like a bug due to that we don't currently fallback to the enterprise URL passed to ARC at https://github.com/actions-runner-controller/actions-runner-controller/blob/dc29e31bcc363898e7977a2867a11dea6530278b/controllers/multi_githubclient.go#L196-L199

The `secretDataToGitHubClientConfig` function assumes the enterprise url is provided via a kv in a K8s secret https://github.com/actions-runner-controller/actions-runner-controller/blob/dc29e31bcc363898e7977a2867a11dea6530278b/controllers/multi_githubclient.go#L283. The problem here is that even if the kv was empty, it doesn't fallback to the controller-wide setting of enterprise url.

If you have some time contributing code by any chance, it would be great if you could add something like:

```
if conf. EnterpriseURL == "" {
    // fallback to the controller-wide setting
    conf.EnterpriseURL = mc.githubClient.EnterpriseURL // Was this possible? I don't remember..
}
```

in L200 or some other place, build your own ARC controller image by running `make docker-buildx docker-push`, and update the image tag in your helm release to the image just pushed.

This should be fixed before 0.26.0 gets released, so even if don't have time, you can expect me to fix it before the release!

Thanks in advance for your support!

**Comment by Jalmeida1994** [comment]
Sure, I'll try and take a look!

First time deep diving the controller code, so sorry in advance, for delay, dumb questions and such.

**Comment by Jalmeida1994** [comment]
I don't know if I should post this on the **Discussions** tab but as it's so close to this I'll start it here:

I noticed that the controller still needs the `controller-manager` secret with GitHub authentication when deploying it. My idea was to only deploy the controller on a system namespace and then use other namespaces to deploy organization runners. As of now I need to add the secret with a dummy GitHub Authentication method just to deploy the controller. What are your thoughts on it? Should it be allowed to deploy the controller without the need for a GitHub Auth method?

**Comment by mumoshu** [comment]
@Jalmeida1994 Hey! Thanks for the fix- It's going to be merged soon.

Regarding the controller-manager seret req, yes, it's currently intended. The explanation on why can be found in the second paragraph of the PR description for https://github.com/actions-runner-controller/actions-runner-controller/pull/1371. TL;DR; There's a feature called "runner groups visibility", which needs to call GitHub API in order to decide which HorizontalRunnerAutoscaler to scale. Obviously, until we decide which HRA to scale, we can't decide which `githubAPICredentialsFrom` and the secret to be used. It's a kind of chicken-and-egg problem.

I think runner groups visibility is quite useful when you're on a large enterprise and wants to use runner groups to control which runner groups to be used for a particular repo, without forcing the user to specify possible complex set of runner labels in their actions workflow definition's `on` field.

If you never use the feature, I think an explicit controller flag to allow omitting the controller-wide GItHub API creds. That would also require a feature to let ARC emit an validation error when it encountered any GitHub API creds secret missing `enterprise_url`, because without the default secret we have no value to fallback.

WDYT?

**Comment by mumoshu** [comment]
@Jalmeida1994 Oops, this has been closed due to the #1725 merge. Let me clearly say that I'm interested in your opinion. If you think it is worth a feature request, I'd appreciate it if you could submit one!

Honestly, I've never managed/operated GHES so I'm willing to hear opinions and feedbacks from a real GHES user like you.

**Comment by Jalmeida1994** [comment]
> @Jalmeida1994 Oops, this has been closed due to the #1725 merge. Let me clearly say that I'm interested in your opinion. If you think it is worth a feature request, I'd appreciate it if you could submit one!

Ahahah no problem, I completely understood! 

> I think runner groups visibility is quite useful when you're on a large enterprise and wants to use runner groups to control which runner groups to be used for a particular repo, without forcing the user to specify possible complex set of runner labels in their actions workflow definition's on field.

Hmm yeah it makes sense seeing that way. You're right, it's better to stay this way. It is never good when we overengineer a solution. The workaround is fairly simple: we just created a dummy authentication method for the default controller deployment secret.

Thank you so much for listening and again for the awesome project. Keep it up. I'll try to test the controller on a GHES environment whenever possible and report back. Cheers!
